co-simulation/fed/mosaic-carma-cloud/src/main/java/org/eclipse/mosaic/fed/carmacloud/ambassador/CarmaCloudRegistrationReceiver.java

```diff
@@ -0,0 +1,139 @@
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import com.google.gson.Gson;
+import java.io.DataInputStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Worker thread Runnable for operating a listen socket to receive outbound
+ * CARMA Cloud Messages from CARMA Cloud instances
+ * This {@link Runnable} instance will operate a UDP socket to subscribe to
+ * packets from a CARMA Cloud adapter. 
+ * Upon receiving a packet, it will be queued for the primary thread
+ * to process the data once it ticks to a simulation processing step
+ */
+public class CarmaCloudRegistrationReceiver implements Runnable
+{
+	private static final int LISTEN_PORT = 1617; // which port for CARMA Cloud?
+	private final AtomicBoolean running = new AtomicBoolean(false);
+	private final Logger log = LoggerFactory.getLogger(this.getClass());
+	private final Queue<CarmaCloudRegistrationMessage> rxQueue = new LinkedList<>();
+	private ServerSocket m_oSrvr;
+
+
+	/**
+	 * Initialize the listen socket for messages from the CARMA Cloud instance adapter
+	 * 
+	 * @throws RuntimeException if socket instantiation fails
+	 */
+	public void init()
+	{
+		try
+		{
+			m_oSrvr = new ServerSocket(LISTEN_PORT);
+		}
+		catch (Exception e)
+		{
+			throw new RuntimeException(e);
+		}
+	}
+
+
+	/**
+	 * The main method of the worker thread. Listens for incoming messages and
+	 * queues them for processing on the primary thread.
+	 */
+	@Override
+	public void run()
+	{
+		running.set(true);
+		try
+		{
+			while (running.get())
+			{
+				Socket oSock = m_oSrvr.accept();
+        DataInputStream oIn = new DataInputStream(oSock.getInputStream());
+
+				// parse message
+				CarmaCloudRegistrationMessage parsedMessage = 
+					new Gson().fromJson(oIn.readUTF(), CarmaCloudRegistrationMessage.class);
+
+				// Enqueue message for processing on main thread
+				synchronized (rxQueue)
+				{
+					log.info("New CARMA Cloud instance '{}' received with CARMA Cloud Registration Receiver.", parsedMessage.getId());
+					rxQueue.add(parsedMessage);
+				}
+			}
+		}
+		catch (Exception oEx)
+		{
+			log.error("Error occurred", oEx);
+		}
+	}
+
+
+	/**
+	 * Stop the runnable instance and close the listen socket.
+	 */
+	public void stop()
+	{
+		running.set(false);
+		if (m_oSrvr != null)
+		{
+			try
+			{
+				m_oSrvr.close();
+			}
+			catch (Exception oEx)
+			{
+				log.error("Error occurred", oEx);
+			}
+		}
+	}
+
+
+	/**
+	 * Query the current buffer of outbound messages. Clears the currently stored
+	 * buffer once called. Thread-safe.
+	 * 
+	 * @return The list of received outbound message from all Infrastructure Device
+	 *         instances since last call of this method
+	 */
+	public List<CarmaCloudRegistrationMessage> getReceivedMessages()
+	{
+		List<CarmaCloudRegistrationMessage> output = new ArrayList<>();
+		synchronized (rxQueue)
+		{
+			output.addAll(rxQueue);
```

Comment:
Nit: I dont think the add all operation needs to be inside the synchronized statement. Only modifications to the rxQueue need to be inside this statement

Reply:
I believe @kruegersp was directly copied from [CARMA Ambassador](https://github.com/usdot-fhwa-stol/cdasim/blob/b7c197777637e88f31e165329180218de543b877/co-simulation/fed/mosaic-carma/src/main/java/org/eclipse/mosaic/fed/carma/ambassador/CarmaRegistrationReceiver.java#L102-L103) which was implemented by Kyle. @kjrush do you think this modification will proceed without any issues?

Reply:
Are we certain the synchronization isn't to prevent rxQueue modification while the addAll operation iterates?

Reply:
Hmm good point bryan. That is probably correct